src/downloader.rs

```diff
@@ -0,0 +1,127 @@
+//! Represents the download controller.
+
+use crate::{download::Download, Error};
+use futures::stream::{self, StreamExt};
+use reqwest_middleware::{ClientBuilder, ClientWithMiddleware};
+use reqwest_retry::{policies::ExponentialBackoff, RetryTransientMiddleware};
+use reqwest_tracing::TracingMiddleware;
+use std::fs;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+use tracing::debug;
+
+const DEFAULT_RETRIES: u32 = 3;
+const DEFAULT_CONCURRENT_DOWNLOADS: usize = 32;
+
+/// Represents the download controller.
+#[derive(Debug, Clone)]
+pub struct Downloader {
+    /// Directory where to store the downloaded files.
+    directory: PathBuf,
+    /// Number of retries per downloaded file.
+    retries: u32,
+    /// Number of maximum concurrent downloads.
+    concurrent_downloads: usize,
+}
+
+impl Downloader {
+    /// Starts the downloads.
+    pub async fn download(&self, downloads: &[Download]) {
+        // Prepare the HTTP client.
+        let retry_policy = ExponentialBackoff::builder().build_with_max_retries(self.retries);
+        let client = ClientBuilder::new(reqwest::Client::new())
+            .with(TracingMiddleware)
+            .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+            .build();
+
+        // Download the files asynchronously.
+        let _tasks = stream::iter(downloads)
+            .map(|d| self.fetch(&client, d))
+            .buffer_unordered(self.concurrent_downloads)
+            .collect::<Vec<_>>()
+            .await;
+    }
+
+    /// Fetches the files and write them to disk.
+    async fn fetch(&self, client: &ClientWithMiddleware, download: &Download) -> Result<(), Error> {
+        // Request the file.
+        debug!("Fetching {}", download.url.clone());
+        let res = client.get(download.url.clone()).send().await.unwrap();
+
+        // Prepare the destination directory/file.
+        fs::create_dir_all(&self.directory)?;
+        let output = self.directory.join(&download.filename);
+        debug!("Creating file {:?}", &output);
+        let mut file = File::create(output)?;
+
+        // Download the file chunk by chunk.
+        debug!("Retrieving chunks...");
+        let mut stream = res.bytes_stream();
+        while let Some(item) = stream.next().await {
+            let chunk = item?;
+            file.write_all(&chunk)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// A builder used to create a [`Downloader`].
+pub struct DownloaderBuilder(Downloader);
+
+impl DownloaderBuilder {
+    /// Creates a builder with the default options.
+    pub fn new() -> Self {
+        DownloaderBuilder::default()
+    }
+
+    /// Sets the directory where to store the [`Download`]s.
+    pub fn directory(mut self, directory: PathBuf) -> Self {
```

Comment:
let's use [Imperative mood](https://en.wikipedia.org/wiki/Imperative_mood) for anything that mutates or retrieves: `fn set_directory(mut self)` or `fn get_directory(&self)`

Reply:
Ah, yes, makes sense. But this convention does not apply here because I am in a builder object.

However after reading your comment and digging a bit, there is actually [a Rust RFC to name setters and getters](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis). It also looks like for mutables, the idiomatic way is to add the `_mut` suffix (https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter).

Reply:
Nice find!